File: apps/mobile-poc/android/app/src/main/java/cn/edu/qlu/toolbox/mobilepoc/GradeExportSecurity.java
package cn.edu.qlu.toolbox.mobilepoc;

import java.net.URI;
import java.util.Set;

final class GradeExportSecurity {
    static final String ACADEMIC_HOST = "jw.qlu.edu.cn";
    static final String SSO_HOST = "sso.qlu.edu.cn";
    private static final Set<String> ALLOWED_HOSTS = Set.of(ACADEMIC_HOST, SSO_HOST);

    private GradeExportSecurity() {}

    static boolean isAllowedUrl(String rawUrl) {
        try {
            URI uri = URI.create(rawUrl);
            int port = uri.getPort();
            return "https".equalsIgnoreCase(uri.getScheme())
                && uri.getHost() != null
                && ALLOWED_HOSTS.contains(uri.getHost().toLowerCase())
                && uri.getUserInfo() == null
                && (port == -1 || port == 443);
        } catch (IllegalArgumentException error) {
            return false;
        }
    }

    static boolean isLoggedInUrl(String rawUrl) {
        if (!isAllowedUrl(rawUrl)) return false;
        URI uri = URI.create(rawUrl);
        if (!ACADEMIC_HOST.equalsIgnoreCase(uri.getHost())) return false;
        String path = uri.getPath() == null ? "" : uri.getPath();
        String query = uri.getQuery() == null ? "" : uri.getQuery();
        return path.contains("/jwglxt/xtgl/index_initMenu.html")
            || path.contains("/jwglxt/cjcx/")
            || (path.startsWith("/jwglxt/") && query.contains("jsdm=xs"));
    }
}

File: apps/mobile/android/app/src/main/java/io/github/c1oudreamw/lumatile/GradeExportSecurity.kt
package io.github.c1oudreamw.lumatile

import java.net.URI

internal object GradeExportSecurity {
    const val ACADEMIC_HOST = "jw.qlu.edu.cn"
    const val SSO_HOST = "sso.qlu.edu.cn"
    private val allowedHosts = setOf(ACADEMIC_HOST, SSO_HOST)

    fun isAllowedUrl(rawUrl: String?): Boolean = try {
        val uri = URI.create(rawUrl ?: return false)
        uri.scheme.equals("https", ignoreCase = true) &&
            uri.host?.lowercase() in allowedHosts &&
            uri.userInfo == null &&
            (uri.port == -1 || uri.port == 443)
    } catch (_: IllegalArgumentException) {
        false
    }

    fun isLoggedInUrl(rawUrl: String?): Boolean {
        if (!isAllowedUrl(rawUrl)) return false
        val uri = URI.create(rawUrl)
        if (!uri.host.equals(ACADEMIC_HOST, ignoreCase = true)) return false
        val path = uri.path.orEmpty()
        val query = uri.query.orEmpty()
        return path.contains("/jwglxt/xtgl/index_initMenu.html") ||
            path.contains("/jwglxt/cjcx/") ||
            (path.startsWith("/jwglxt/") && query.contains("jsdm=xs"))
    }

    fun blockedHost(rawUrl: String?): String = try {
        URI.create(rawUrl.orEmpty()).host?.takeIf { it.isNotBlank() } ?: "未知地址"
    } catch (_: IllegalArgumentException) {
        "未知地址"
    }
}
